Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,484 @@
+    }
+
+    return alternatives;
+}
+
+bool MailWidget::is_supported_alternative(Alternative const& alternative) const
+{
+    auto lowercase_type = alternative.body_structure.type.to_lowercase();
+    auto lowercase_subtype = alternative.body_structure.subtype.to_lowercase();
+    return lowercase_type == "text" && lowercase_subtype.is_one_of("plain", "html");
+}
+
+void MailWidget::selected_mailbox()
+{
+    const auto& index = m_mailbox_list->selection().first();
+
+    if (!index.is_valid())
+        return;
+
+    auto& base_node = *static_cast<BaseNode*>(index.internal_data());
+
+    if (is<AccountNode>(base_node)) {
+        // FIXME: Do something when clicking on an account node.
+        return;
+    }
+
+    auto& mailbox_node = verify_cast<MailboxNode>(base_node);
+    auto& mailbox = mailbox_node.mailbox();
+
+    VERIFY(!(mailbox.flags & (unsigned)IMAP::MailboxFlag::NoSelect));
+
+    auto response = m_imap_client->select(mailbox.name)->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to select mailbox. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to select mailbox. The server says: '{}'", response.response_text()));
+        return;
+    }
+
+    Vector<IMAP::SearchKey> keys;
+    // clang-format off
+    keys.append(IMAP::SearchKey {
+        IMAP::SearchKey::All {}
+    });
+    // clang-format on
+    response = m_imap_client->search({}, move(keys), false)->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to retrieve e-mails. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve e-mails. The server says: '{}'", response.response_text()));
+        return;
+    }
+
+    auto search_results = move(response.data().search_results());
+    if (search_results.is_empty()) {
+        m_individual_mailbox_view->set_model(InboxModel::create({}));
+        return;
+    }
+
+    // clang-format off
+    auto fetch_command = IMAP::FetchCommand {
+        .sequence_set = { { (int)search_results.first(), (int)search_results.last() } },
+        .data_items = {
+            IMAP::FetchCommand::DataItem {
+                .type = IMAP::FetchCommand::DataItemType::BodySection,
+                .section = IMAP::FetchCommand::DataItem::Section {
+                    .type = IMAP::FetchCommand::DataItem::SectionType::HeaderFields,
+                    .headers = { { "Subject" } }
+                }
+            },
+            IMAP::FetchCommand::DataItem {
+                .type = IMAP::FetchCommand::DataItemType::BodySection,
+                .section = IMAP::FetchCommand::DataItem::Section {
+                    .type = IMAP::FetchCommand::DataItem::SectionType::HeaderFields,
+                    .headers = { { "From" } }
+                }
+            }
+        }
```

Comment:
You can combine these two into a single DataItem which asks for both the subject and from fields.

Reply:
Oh nice, fixed.